### DISCUSSION TITLE
[@types/sinon] Replace neither takes nor recieves SinonSpy type

### DISCUSSION THREAD

**Discussion by michaeljones** [DefinitelyTyped/DefinitelyTyped#55049] [Issues with a @types package]
My understanding is that it is common to call: 

```javascript
const fake = sinon.fake.returns('23')
sinon.replace(module, "function", fake)
```
or
```javascript
const fake = sinon.replace(module, "function", sinon.fake.returns('23'))
```
But given:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/248cee77c9a96d4aa594d15b0fbc3bffe49b48dd/types/sinon/index.d.ts#L1610-L1615

`sinon.replace` is not set up to either receive or return a sinon fake or spy (the `sinon.SinonSpy` type) so it is hard to called replace using a fake and if using the return value of `sinon.replace` it is hard to access "spy" methods or values such as "callCount" or "firstCall". https://sinonjs.org/releases/v11.1.2/spy-call/

The `sinon.replace` types are instead set up to focus on the type of the thing being replaced rather than what we're replacing it with. I can see the benefits of this to some degree but it would be great if they could be mixed some how. Not sure if that is possible though.

[a comment by an automated account is not shown]

**Comment by 43081j** [comment]
You're correct, the docs say it can take a fake, a spy or a stub. So we should type it the way it is but have overloads possibly which take those types (but still strongly typed to the correct parameters/return types etc)
